### DISCUSSION TITLE
Support multiple master key configurations per database

### DISCUSSION THREAD

**Discussion by frankmorgner** [keepassxreboot/keepassxc#3597] [Ideas]
##  Summary
When creating (or unlocking) a DB, keepassxc allows specifying what's known as composite key in keypass2, e.g. a master password *with* a key file. *All* those credentials seem to be required when unlocking the DB.

## Desired Behavior
Ideally, I'd like to use the the key file if it is available. If it is not available, I'd like to use the master password. So the question is whether it is possible to unlock the DB with a single credentials from many possible credentials. Would that be possible or is this already implemented?

## Possible Solution
In essence, we're not creating a composite key anymore, instead we're creating a key from each credential which independently unlock the DB.

## Context
Think of the way Pluggable Authentication Modules (PAM) can be stacked, combined or used separately, for example.

**Comment by droidmonkey** [comment]
This is not how encryption works.

**Comment by frankmorgner** [comment]
OMG, I hope you're not treating all your users like this. Let's just say we had a bad start and I'll explain with a different example:

When using disk encryption with multiple user accounts you have a single password protected key for each user. This user key is then used to decrypt the disk encryption key from a user specific cipher text so that every user can decrypt the same disk without knowing anything from the other users. This is not a problem of encryption, just a matter of key management. Would it be possible to add such key management to the Keepass DB?

**Comment by droidmonkey** [comment]
Ok that makes more sense, you store multiple encrypted versions of the master key with the various credentials. Here is my personal issue with this from a database perspective. If you are able to decrypt with JUST the password, your key file is totally  irrelevant and does not provide any additional protection. So just don't use it. 

Sorry for the blunt response.

**Reply by michaelblyons** [reply]
I agree that the example provided is bad, but once (more) hardware tokens are working, having two approved hardware tokens makes sense to me, lest one be lost.

**Comment by frankmorgner** [comment]
Just be more creative: What about password+Challenge/Response that can be used as alternative to smart card+PIN? Both variants have a similar security level allowing me to use whatever I am carrying with me...

**Comment by frankmorgner** [comment]
According to [the documentation, KDBX 4 at least has some room to store custum data for plugins](https://keepass.info/help/kb/kdbx_4.html). That could be a fit for alternative decryption credentials...

**Comment by droidmonkey** [comment]
Where this does lower security (especially if you allow password-only in lieu of Challenge-Response/Key File) it also allows for multiple users of the same database with different passwords (also "dead man switch"). There is some value in supporting this feature, but it would be KeePassXC only supported (not portable) and carries risk that must be communicated to the user. This also complicates the master key setup.

**Comment by frankmorgner** [comment]
What's the status of interoperability? Which KDBX features are supported across different implementations?

**Comment by droidmonkey** [comment]
We are fully interoperable with KeePass 2.x, all active clones, and several KeePass plugins. The only KeePass KDBX feature we don't support yet is tags. Supporting multiple credentials would certainly be an outlier in that model.

**Comment by droidmonkey** [comment]
After storyboarding this for a little bit I came up with a couple gotchas:

1. Anytime the original master key changes (manually or when using Yubikey) then all additional keys would need to be entered to encrypt the new derived master key.
2. When entering credentials, we would have to assume it was the original master key first, fail decryption, then test any additional keys. Given a KDF that takes 1 second to derive, this could lead to multiple seconds of delay (even more so on lesser powered devices).
3. As discussed above, you can inadvertently lower your security threshold with additional credentials that are less robust.

This discussion is very similar to the "give the FBI a master key" concept. Having multiple pathways to an encrypted resource ultimately reduces the security of that resource.

![image](https://user-images.githubusercontent.com/2809491/66270107-ad422680-e81d-11e9-9df2-bebc337140b1.png)

**Reply by jokesper** [reply]
> then all additional keys would need to be entered

I am probably missing something, but I think this could be solved.

We could store them in the decrypted database.
If I understand the decryption via yubikeys correctly,
we would need to store the challenge response in the decrypted database
and can only rotate the challenge for the connected keys.
The salt can still be rotated, since we can derive the new master keys via the stored secrets.

This does come at the cost of some security, though I think this may less important (A breached database leaking all decryption passwords, e.g. Breach via weak alternative password would leak a strong primary password).
With a proper warning it should be ok.

> Given a KDF that takes 1 second to derive, this could lead to multiple seconds of delay

This could be solved by allowing a drop down of which key from the header to try,
but this kind of breaks the zero-knowledge decryption.
Though if we don't pad with lots of random fake keys (and we shouldn't, this increases decryption failure times)
this information would be unencrypted in the headers anyway.

**Reply by amluto** [reply]
> After storyboarding this for a little bit I came up with a couple gotchas:
> 
> 1. Anytime the original master key changes (manually or when using Yubikey) then all additional keys would need to be entered to encrypt the new derived master key.
> 2. When entering credentials, we would have to assume it was the original master key first, fail decryption, then test any additional keys. Given a KDF that takes 1 second to derive, this could lead to multiple seconds of delay (even more so on lesser powered devices).
> 3. As discussed above, you can inadvertently lower your security threshold with additional credentials that are less robust.
> 
> This discussion is very similar to the "give the FBI a master key" concept. Having multiple pathways to an encrypted resource ultimately reduces the security of that resource.
> 
> ![image](https://user-images.githubusercontent.com/2809491/66270107-ad422680-e81d-11e9-9df2-bebc337140b1.png)

It might break all kinds of compatibility, but I think these problems are solvable for real without any particular user friction.  Instead of wrapping the master key with symmetric cryptography, the wrapping slots could all use asymmetric cryptography.  Traditionally this would be public-key encrypt and decrypt operations, but the modern paradigm would be to use a KEM (which has logically essentially the same effect).

So slot 1 could be a password (or a random code that the user writes down somewhere safe), and the password would be used to derive a key pair.  Slot 2 could be a key pair where the secret key is wrapped by a Yubikey or TPM, or it could even be set up so that a hardware device and a short password are needed to decrypt it.  (And enough extra data could be stored so that the keypair could be regenerated and re-wrapped without access to the hardware device - TPM definitely supports this, and at least FIDO1 can be fudged a bit to pull this off.)  Whenever a master key is (re)generated, a KEM is run against each slot's public key and the master key is wrapped by the resulting slot-specific secret.  All the wrapped copies would be stored.

The result would be that the user could save the database, and the save process could very quickly generate a new master key, compute all the wrapped copies, and write them to disk.  And the user could subsequently decrypt the database using any of the slots' secrets.

If hardware tokens were wired up very nicely, even an prior decryption session using a token that was fully captured by an attacker could not be used to decrypt a newer database.

**Comment by mxsrc** [comment]
I like this proposal a lot, I think there are use-cases where this would not lower security, but rather improve it.

Consider a user who secures the database using only a strong password, rather than a weaker password with a keyfile, such that the database is self-contained and can be unlocked just with the password. If he'd like to automatically unlock the database when logging into/unlocking the computer, he might be tempted to use a simpler, less secure password, since it has to be typed regularly. With the proposed feature, it might be possible to have two variants of accessing the database: with a strong password, or with a weaker password and a keyfile that is present only on the computer with the associated login.

**Reply by speakman** [reply]
This is exactly what I'm looking for. Or just that simple case of using multiple Yubikeys - current FAQ _even recommend_ you to use two Yubikeys _using the same key_!!

In my case I would like to either have a _very strong_ password (or rather a passphrase, see xkcd 936) OR a Yubikey. Where the passphrase is the backup if the yubikey gets lost.

LUKS is using "slots" where different methods can be combined. Any active key will unlock the master key which in turn is used to encrypt the disk.

For a bigger company, handing out multiple Yubikeys is a common practice. Whenever a keys is lost, it's keys/certs are revoked immediately but there are no disturbance in the work for the employee. A new backup key is sent out. Just thinking of having SAME KEY in these keys is truly a nightmare!

**Reply by frankmorgner** [reply]
yeah, this advice is rather stupid.

I've originally posted this here, because I thought keypassxc was a bit more permissive ragarding new ideas. Anyway, I switched back to KeyPass for using https://github.com/mlaily/KeePass-CertificateShortcutProvider. This creates a smartcard (yubikey) encrypted blob on your disk containing your password. Basically, your password is still the only key to your database, but now you can decide whether you want to type the password or if you want to use your token to recover the password instead.

**Reply by frankmorgner** [reply]
This approach should work with any number of tokens...

**Reply by amluto** [reply]
> I like this proposal a lot, I think there are use-cases where this would not lower security, but rather improve it.
> 
> Consider a user who secures the database using only a strong password, rather than a weaker password with a keyfile, such that the database is self-contained and can be unlocked just with the password. If he'd like to automatically unlock the database when logging into/unlocking the computer, he might be tempted to use a simpler, less secure password, since it has to be typed regularly. With the proposed feature, it might be possible to have two variants of accessing the database: with a strong password, or with a weaker password and a keyfile that is present only on the computer with the associated login.

There's also the usecase of a user who is concerned that their computer might be compromised at some point and doesn't want to type in their strong password (or backup code or whatever) since it might get keylogged.  But they still want it as a backup.

**Comment by d-faure** [comment]
IMHO, if the primary reason to a tool such as KeepassXC is the user's inability to remember it's passwords, why do believe that he can have better chances at this with the master one? Therefore an alternate authentication feature should be available for the master password to let him have a chance if he needs to.

An interesting implementation could for example take advantage of [Shamir's secret sharing algorithm](https://en.wikipedia.org/wiki/Shamir%27s_Secret_Sharing) in order to securely take account of a subset of user conditions to validate the database opening (the regular master password, the key file or even the yubikey token being some of them).

**Reply by d-faure** [reply]
My example algorithm wasn't the right one to use.
Cf. [this answer](https://stackoverflow.com/a/597194/4153864) to [encryption/decryption with multiple keys](https://stackoverflow.com/questions/597188/encryption-decryption-with-multiple-keys), thre's perhaps better way to handle this.

**Comment by cornfeedhobo** [comment]
It's really a shame that this never went anywhere. LUKS was right to have slots and time has proven that this does not weaken the net security model of the user, especially given that the average user won't use this feature and thus the attack surface remains the same.

If someone decides to tackle this, I'd be willing to pay a bounty.

**Comment by iameru** [comment]
I would also enjoy this feature a lot.
This could also be used in having a somewhat friendly password rotating system. currently, changing the master key just means I will be locked out if I cant remember the new one which i just got and dont have memorized yet. Having multiple slots I could have the old one as a fallback for some days until turning it off.

**Comment by Hoeze** [comment]
I just got a new Yubikey and wanted to set this up:
- long passphrase as backup
- short passphrase + yubikey for everyday use

I was catched by surprised that KeePassXC does not support this and would love to use this feature :+1:

**Comment by Hoeze** [comment]
Adding another use case:
In our lab, we have a KeePassXC database for all important passwords shared between us admins.
My plan was to add each admin's yubikey as one unlock token instead of sharing a single password between us.
This is only possible with key management.

**Reply by michaelk83** [reply]
If you're sharing the database with all the important passwords, then using a separate password for each admin doesn't really add much security, as I understand. If any one of them is compromised, all the passwords can be accessed.

But, you can set this up using helper databases and [AutoOpen](https://keepassxc.org/docs/KeePassXC_UserGuide#_automatic_database_opening). The main DB uses the long secure passphrase, and each helper DB is set up to use one of the YubiKeys, and to AutoOpen the main DB.

**Reply by Hoeze** [reply]
Oh cool, I did not know that you can use AutoOpen to open other databases!
Thanks for the tip :)

---
Yes, it is not really more secure if the DB gets exported when unlocked.
However, hardware tokens cannot be cloned and add some convenience if I just need to press a button on my yubikey + type some short PW.
Also, since we rotate the DB at least yearly, I do not need to remember changing passphrases.

**Reply by droidmonkey** [reply]
@Hoeze this example is a good time to note that KeePassXC is not an enterprise credential management solution. KeePass in general is fundamentally a personal/individual credential manager. You can do limited team access, but don't abuse the flexibility 😅

**Comment by HexF** [comment]
This post is kind of long, but describes a system which should solve this problem, and addresses and summarizes a few of the previous points.


I did some playing around with this idea, and came up with the a system which I believe solves this problem well, remaining fully backwards compatible with the existing master key system.
This diagrams shows the general layout for how keys are utilised to eventually decrypt the database.
Each box represents a piece of information, where green boxes represent those of publicly stored information and red are user-provided secrets (e.g. passwords).
Links between boxes describe operations to derive information.

<img width="1467" alt="Screenshot 2025-04-27 at 5 20 46 pm" src="https://github.com/user-attachments/assets/7726a6f4-0519-4b94-89a4-e3b03a9817f7" />


To setup the system:
1. Generate a public/private key pair
2. Derive a combined key from the credentials, and encrypt the private key using this key
3. Generate a sealing key, and use this to encrypt the derived master key of the database
4. Encrypt the sealing key with public key
5. Store a tuple (KDF input parameters, public key, encrypted private key, encrypted sealing key, encrypted master key) in the database headers

To re-key the derived master key (i.e. the main password is changed):
1. Generate a new sealing key, and encrypt the derived master key
2. Encrypt the new sealing key with the public key
3. Store (KDF input parameters, public key, encrypted private key, encrypted **new** sealing key, encrypted **new** master key) in database headers, replacing old entry.

To derive master key (e.g. for encrypting/decrypting database)
1. Perform KDF to get combined key (as done currently)
2. Decrypt private key with combined key
3. Decrypt sealing key with private key
4. Decrypt master key with sealing key

To re-key an additional key:
1. Derive master key using old key (or generate a new one)
2. Re-run setup procedure, overwriting the tuple instead of storing a new one

Ideally we should generate a new master-key for every re-key event, as the key could still be derived from an older version of the database.
We could do this by taking the existing password/keyfile, and just generating a new salt for the KDF.


I would propose using a combination of RSA and CRYSTAL-Kyber, for similar reasons that [Signal](https://signal.org/blog/pqxdh/) does this, i.e. RSA isn't quantum-safe, and further research could prove Kyber to be insecure.
[Harvest now, decrypt later](https://en.wikipedia.org/wiki/Harvest_now,_decrypt_later) could be a real threat to a password database.

The sealing key should be unique per public/private key pair to ensure we don't run into a situation where [having two of the same message encrypted with the same key, can lead to information disclosure](https://crypto.stackexchange.com/questions/1614/rsa-cracking-the-same-message-is-sent-to-two-different-people-problem).
I would propose using a one-time-pad for the sealing key as we only need to seal 32-bytes of information for an AES-256 key.
[One-time-pad is also proven to have perfect secrecy](https://en.wikipedia.org/wiki/One-time_pad#Perfect_secrecy), so this step will add no additional weakness to the crypto system. 

For encrypting the private key, I propose we utilise the same algorithm as the database to reduce implementation complexity.

I also propose we have per-key algorithm selection, incase we need to substitute any of these algorithms out, or can provide something more secure later down the line (similar to how we can change the database encryption and KDF functions).
This would allow for a gradual re-key of a database if needed.


**To address some points previously:**

Noted in https://github.com/keepassxreboot/keepassxc/discussions/3597#discussioncomment-200317, there would be non-portability in this implementation, which is true.
However, this proposed system remains backwards compatible with the existing KDBX keying scheme as to not break compatibility. Other Keepass clones could look to KeepassXC to implement this feature.

Point (1) noted in https://github.com/keepassxreboot/keepassxc/discussions/3597#discussioncomment-200320, where all credentials would be needed to perform a re-key has been solved through the addition of asymmetric keys to the system
To ensure security however as noted before, we should re-key the master key every time we revoke/change a key.
This may require knowing the normal database master password. To get around this we may consider storing the master password in the database.

Point (2) noted in https://github.com/keepassxreboot/keepassxc/discussions/3597#discussioncomment-200320, where we would need to check against every credential, could be solved by having a selection on the unlock screen for selecting which key to login with.

Point (3) noted in https://github.com/keepassxreboot/keepassxc/discussions/3597#discussioncomment-200320, where security could be reduced. 
I think the pros outweigh the cons here, as noted before, some users have weaker master passwords than they usually would because they need to enter this so frequently.
Using a very strong master password (or even an offline keyfile, e.g. stored on a flash drive in a vault), and a combination of a YubiKey/other hardware-backed key storage and a weaker password which a user would want to entire many times over, is a use case I can foresee.
While it may be possible to lower the entire security of the database by having an additional single weak password, I think this should be noted in big warning in the docs before using this feature.
I think the saying 'a chain is only as strong as its weakest link' perfectly summarises this.

https://github.com/keepassxreboot/keepassxc/discussions/3597#discussioncomment-1601960 suggested the use of SSS, which could be good if wanted to require n of m credentials to be present, but I can't see a good use case for this.
However, this system could easily be adapted to implement this through instead of directly encrypting the master key, we encrypt our SSS shares for the configured system.

The system described above works similar this comment https://github.com/keepassxreboot/keepassxc/discussions/3597#discussioncomment-1615083.

**Some other thoughts:**

Currently in the Quick Unlock module, we are doing a very similar thing to this, for example the macOS TouchID module [stores a sealing key with the secure system key storage, which seals the master key](https://github.com/keepassxreboot/keepassxc/blob/f53c7e5af54974c332b460e36cbe2c2bd0a34b12/src/quickunlock/TouchID.mm#L94).
One of my annoyances with the quick unlock system is I need to re-enter my master password every time I restart KeepassXC - which I think could be stored more securely as a key in my laptops Secure Enclave, rather than a weaker password in my head.

My thought is we could have each module in the quick unlock system be a credentials which make up a combined key, instead of directly passing the master key around the codebase.
This would decrease the possibility of developer-error in terms of miss-handling of the master key.
This also allows us to have the option of Quick Unlock modules persisting across KeepassXC restarts, as we could store any parameters they require in the database headers.
In cases where we don't want Quick Unlock modules to persist restarts, (e.g. a PIN module), we can construct the additional key tuples, but flag them as in-memory only, and to not be saved.

This once again, has the incompatibility issue with other Keepass implementations, but the normal master password should still work perfectly fine as fallback. Other implementations could follow our methods if they desire.


I might turn this into a proper Feature Request at some point to get further comments around the cryptosystem design before it gets implemented. The Quick Unlock stuff might be work a follow-on feature request.
As this directly affects database security, I would expect lots of in-depth analysis of potential weaknesses in this system before this makes its way into KeepassXC.

**Reply by droidmonkey** [reply]
You are missing a very important step in the KDBX security model, the rotation of the master seed on every save operation. This changes the "derived master key" for the database and would completely break the lower (or upper, depending on which flow you used to decrypt the database) portion of your flow.

This fact is the reason why we require a yubikey to be connected on save (master seed == yk challenge) and it is also the reason why my original proposal three years ago would require entering all the credential combos in on save.

LUKS works fine because the _master key never changes_ so wrapping the master key with different encryption boxes is totally fine. That is absolutely not the case for KDBX databases. KDBX also uses zero-knowledge decryption, you have _absolutely no idea_ if the database key requires password, key file, yubikey, or any combination. Adding knowledge to that process breaks down the security model substantially (in comparison to zero knowledge).

Honestly I have near zero appetite for this scheme and would likely never incorporate such a complex (and non-standard) change into our application.

In other words, spend your development time on improving the user interface and general workflow of the app and ignore requests for non-standard features.
